### ISSUE TITLE
`dagger do` should check for a valid plan

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#1711]
## Problem

Currently, `dagger do` (and its 0.1 predecessor `dagger up`) does not check that a plan is valid before running it. It delegates this task to the CUE loader, which in turn requires the developer to explicitly apply a top-level definition, like so:

```
import (
  "dagger.io/dagger"
)

dagger.#Plan & {
  // Plan configuration goes here
  ...
}
```

This is valid CUE, and we should support it. But it does not solve several problems caused by the absence of an explicit validity check:

1. Developers can shoot themselves in the foot by declaring fields outside the `dagger.#Plan` definition. For example:

```
dagger.#Plan & {
  ...
}
// Typo in a field name. Will not be caught.
clients: { ... }
// Still using a deprecated field name. Will not be caught.
inputs: { ... }
```

This is normal CUE behavior, but may not be immediately obvious to all developers.

2. If a plan spans multiple files, the top-level incantation must be repeated in each file. This is verbose (which makes DX incrementally worse) and also increases risks of mistakes.

3. It closes the door to denying top-level configuration fields. We're not yet certain which pattern is best: allow unrelated top-level fields to co-exist with the plan? Or forbid it? In doubt, we should go for the most reversible choice, which is to deny it (we can always allow it later with no breakage, but the opposite is not true). This would be a side effect of checking for valid plans.

## Solution

Modify `dagger do` to explicitly check for a valid plan by unifying it against `dagger.#Plan`.

**Comment by shykes** [comment]
Any takers @dagger/maintainers ? This one is time sensitive, since it makes schema enforcement more strict (no top-level fields allowed if they're not part of the plan definition). Once merged, we have more options.

**Comment by jlongtine** [comment]
@shykes: I had a good chat with @aluzzardi about this. 

First, I know how to accomplish this. It's pretty straightforward (finally figured it out):

```golang
// This loads the #Plan schema as a cue.Value
daggerInstances := cueload.Instances([]string{pkg.DaggerPackage}, &cueload.Config{Dir: src})
daggerInst := daggerInstances[0]

daggerValue := c.Context.BuildInstance(daggerInst)
planSchema := daggerValue.LookupPath(cue.MakePath(cue.Def("#Plan")))

// This is the user's CUE config
v, err := c.Context.BuildInstances(instances)

// Now we Unify the user's CUE config against the #Plan schema:
val := v[0]
val = val.Unify(planSchema)
```

As an aside, the way to accomplish a similar loading of a package via the CUE cli is: `cue eval dagger.io/dagger`

-----

Now for a few questions about the intent and potential side effects of various implementations...

### Different behavior: `dagger` vs `cue eval`

One concern that @aluzzardi and I have is that enforcement of the `dagger.#Plan` schema on the root `cue.Value` within `dagger` will mean that unless our users use this form:
```
dagger.#Plan & {
  // Plan configuration goes here
  ...
}
```
the output of `cue def`/`cue eval` won't behave like dagger does, cutting off a potential debug utility. 

### Backwards incompatible change/degradation of DX

Another concern we have is that enforcing `dagger.#Plan` will mean that patterns like this one (`#PythonBuild`) in [Building docker container images](https://docs.dagger.io/1205/container-images/) (one of our Core Concepts docs), will no longer work, because we can't add new fields -- including definitions --  to `dagger.#Plan`. Related, it prevents `dagger` users from have any non-`dagger` CUE within their plan (excepting subfields of `actions`, because of the open definition of `actions`).

I've also seen this pattern in some code from @lyzhang1999 (so this change would break their plans). 

I can see how this might be seen as a benefit as it forces users to create "external" packages (either in `cue.mod/usr` or as sub-packages in the root module) for patterns they are trying to create. But it also reduces the ability for users to incrementally move toward that, if they want.

### Other possible solutions

1. We could warn on fields that are close to (Levenshtein distance or other similarity function) fields in `dagger.#Plan`, to reduce the likelihood of mistyped field names (including old deprecated field names like `inputs`)
2. We could have `dagger project init` create a nice scaffold hello world example (perhaps with `--scaffold`) that provides the `dagger.#Plan & { ... }` form out of the box. 

### Conclusions and next steps

All in all, I'm not sure how to balance the various DX concerns here. I understand the desire to reduce the types of errors and mistakes that you've outlined, but I'm also seeing the potential for other sorts of errors and DX degradation (including a pattern we've described in our docs) from occurring by implementing this solution. 

All in all, after reviewing this, I think we've already past the point of being able to enforce this without user impact, so we should probably tread more carefully. And it has me wondering if we should find other ways to reduce the possibility of user error via misspelled fields (or the use of deprecated fields).

**Comment by shykes** [comment]
> First, I know how to accomplish this. It's pretty straightforward (finally figured it out):

Great news!

> ### Different behavior: `dagger` vs `cue eval`

> One concern that @aluzzardi and I have is that enforcement of the `dagger.#Plan` schema on the root `cue.Value` within `dagger` will mean that unless our users use this form:
> [...]
> the output of `cue def`/`cue eval` won't behave like dagger does, cutting off a potential debug utility.

That's true but 1) long term there's a path to solving this problem cleanly, and 2) in the short term, it doesn't seem to be a big issue.

* Long term path: there really ought to be a way to unify the root value with a definition, from the command line. Even separately from Dagger, I have felt the need for this feature, and been frustrated that it didn't exist. For example, to quickly vet a configuration against a particular schema, without having to explicitly edit that configuration. Something like `cue --unify-root`. We can contribute this feature upstream. And as a step towards contributing it upstream, we can add it to our `dagger cue` wrapper command (which we need to ship sooner rather than later).

* Short term: it doesn't seem to be a pressing issue short term. I haven't felt the need to use `cue eval` for debugging in a while, except for very basic things like checking the value of a field. I'm not saying there is zero need for it, but it seems pretty low so far.

All in all, there is potentially a slight DX degradation, but there's a path to fixing it, and the DX gain - reducing confusion for all beginner - is much larger. So I think it's a good tradeoff.

> ### Backwards incompatible change/degradation of DX
> Another concern we have is that enforcing `dagger.#Plan` will mean that patterns like this one (`#PythonBuild`) in [Building docker container images](https://docs.dagger.io/1205/container-images/) (one of our Core Concepts docs), will no longer work, because we can't add new fields -- including definitions -- to `dagger.#Plan`.

Two different things here:

* Top-level *definitions* like `#PythonBuild` will continue to work, because closedness does not apply to them
* Top-level hidden fields will also continue to work, for the same reason
* Top-level regular fields will break. The sooner we merge this change, the less developers will be affected by this. Note that there is an easy fix, which is to change top-level fields to hidden fields.

> Related, it prevents `dagger` users from have any non-`dagger` CUE within their plan (excepting subfields of `actions`, because of the open definition of `actions`).
> I've also seen this pattern in some code from @lyzhang1999 (so this change would break their plans).
> 
> I can see how this might be seen as a benefit as it forces users to create "external" packages (either in `cue.mod/usr` or as sub-packages in the root module) for patterns they are trying to create. But it also reduces the ability for users to incrementally move toward that, if they want.

Yes, that's correct. This pattern of embedding a Dagger plan within an existing configuration, deserves further discussion. I don't feel like we know enough about usage patterns to lock ourselves in a direction. Until we do: "no is temporary, yes is forever". With this course of action we keep both options open.

> ### Other possible solutions
> 1. We could warn on fields that are close to (Levenshtein distance or other similarity function) fields in `dagger.#Plan`, to reduce the likelihood of mistyped field names (including old deprecated field names like `inputs`)
> 2. We could have `dagger project init` create a nice scaffold hello world example (perhaps with `--scaffold`) that provides the `dagger.#Plan & { ... }` form out of the box.

I still like the original option better :)

> ### Conclusions and next steps
> All in all, I'm not sure how to balance the various DX concerns here. I understand the desire to reduce the types of errors and mistakes that you've outlined, but I'm also seeing the potential for other sorts of errors and DX degradation (including a pattern we've described in our docs) from occurring by implementing this solution.
> 
> All in all, after reviewing this, I think we've already past the point of being able to enforce this without user impact, so we should probably tread more carefully. And it has me wondering if we should find other ways to reduce the possibility of user error via misspelled fields (or the use of deprecated fields).

Thanks for the input. A few corrections (also mentioned above):

* The pattern in the docs will not break
* Some user configurations will break but not a lot, and the fix is easy
* Learning curve will be improved which is a major pain point right now based on user feedback
* The longer we wait, the more painful this change will be

**Comment by jlongtine** [comment]
@shykes Unfortunately, this is not true:

> Top-level definitions like #PythonBuild will continue to work, because closedness does not apply to them. Top-level hidden fields will also continue to work, for the same reason

As discussed [here](https://github.com/dagger/dagger/issues/473) and [here](https://github.com/cue-lang/cue/issues/543), new definitions are **_not allowed_** to be added to already closed structs. I also have code that demonstrates this. The only way to add fields is by using hidden fields (or definitions). 

Given the reversibility of this decision, I feel fairly good about doing it, with the caveats I listed.

If you're ok with the fact that _no new **fields** or **definitions** of any kind_  can be added to a CUE file that is a `dagger.#Plan`, I can probably finish this up tomorrow.

**Comment by shykes** [comment]
That surprises me, I tried it with cue and it seems to work?

```
$ cue eval - <<-EOF
> #Plan: {
>   hello: string
> }
> 
> #Plan & {
>  #foo: "bar"
> }
> EOF
#Plan: {
    hello: string
}
#foo:  "bar"
hello: string
```

Is my version of CUE too old? It's dev version, commit 2362858826cd0a63fd33691928665c60c3411361 .

**Comment by shykes** [comment]
I just tried again with the latest cue master (commit a03817fce72e220155022bf4fa9b2f68e35aa6b5) and it works. What am I missing?

**Comment by jlongtine** [comment]
I can dig into this more tomorrow, I know it's not working in the version of cue we're currently importing into dagger (which @aluzzardi and I have discussed upgrading). 

There was a _partial_ revert of that change, I believe. So... maybe it's still in our version of `cue`, but not in latest `master`.

**Comment by shykes** [comment]
The first version I tried (2362858826cd0a63fd33691928665c60c3411361) was from July 2021. Fun fact: I was still pulling from the old CUE repo controlled by Google... So that commit is the last commit to master on that repo ;)

**Comment by jlongtine** [comment]
This depends on #2182
